### PR TITLE
fix broadcasting into buffers

### PR DESCRIPTION
**Pull request by lxvm**
Hi,

I'm using Zygote as an AD backend in Integrals.jl and while I was writing tests I noticed I couldn't assign a number to length-1 `Buffer` using broadcasting. I think this is because the method signature for the pullback on `materialize!` is too restrictive, since `copyto!` allows for arbitrary iterators on the rhs of `buf .= itr`. I also added a test for a MWE.

### PR Checklist

- [X] Tests are added
- [X] Documentation, if applicable

Update: A more complete MWE brings up a second issue:
<details>
<summary>MWE 2</summary>

```julia
julia> using Zygote
julia> gradient(1) do p
           b = Zygote.Buffer([1,2,3])
           b .= p
           return sum(copy(b))
         end
ERROR: MethodError: no method matching (::ChainRulesCore.ProjectTo{Float64, @NamedTuple{}})(::Vector{Float64})

Closest candidates are:
  (::ChainRulesCore.ProjectTo{T})(::ChainRulesCore.NotImplemented) where T
   @ ChainRulesCore ~/.julia/packages/ChainRulesCore/zoCjl/src/projection.jl:121
  (::ChainRulesCore.ProjectTo{<:Number})(::ChainRulesCore.Tangent{<:Complex})
   @ ChainRulesCore ~/.julia/packages/ChainRulesCore/zoCjl/src/projection.jl:192
  (::ChainRulesCore.ProjectTo{<:Number})(::ChainRulesCore.Tangent{<:Number})
   @ ChainRulesCore ~/.julia/packages/ChainRulesCore/zoCjl/src/projection.jl:193
  ...

Stacktrace:
 [1] _project
   @ Zygote ~/.julia/dev/Zygote/src/compiler/chainrules.jl:189 [inlined]
 [2] map(f::typeof(Zygote._project), t::Tuple{Int64}, s::Tuple{Vector{Float64}})
   @ Base ./tuple.jl:318
 [3] gradient(::Function, ::Int64, ::Vararg{Int64})
   @ Zygote ~/.julia/dev/Zygote/src/compiler/interface.jl:98
 [4] top-level scope
   @ REPL[11]:1
```
</details>

Update 2: I added an adjoint for copyto! that fixes the MWE, however I'll try to add a generic adjoint for `copyto!(buffer, itr)` next

Update 3: I started about this the wrong way and the manual has [details here](https://docs.julialang.org/en/v1/manual/interfaces/#man-interfaces-broadcasting) on how to bypass broadcasting machinery, so an adjoint for `Base.materialize!` will have to be discarded and has to be replaced by an adjoint for `copyto!(buffer, broadcasted)`

Update 4: I finished writing an adjoint and added a test for broadcasted assignment to a buffer from a generator. I'll happily incorporate any feedback and improvements.

### CODE REVIEW

**Review by darsnack — CHANGES_REQUESTED** [review]
We discussed this PR on call. It looks mostly good, but there are some outstanding points:
- remove extraneous comments
- name anonymous functions
- can you share the error that occurs when `@non_differentiable Base.Broadcast.Broadcasted(::Nothing)` is removed?

**Comment by ToucheSir** [comment]
> I also wanted to ask, since I removed `@adjoint! copyto!(x::Buffer, y::Any)` if we should restore a generic method, and if that should use indexing or iteration to get the elements of `y`?

That would be best. What worries me more is that removing that `copyto!` rule didn't lead to any failures. We probably need to add in tests to fill that gap. As for indexing vs iteration, the rule probably needs to rely on iteration because the [`copyto!` overload for `Buffer`](https://github.com/FluxML/Zygote.jl/blob/54f1e807d5c098a6100c422b54d14f8cd85e0b3c/src/tools/buffer.jl#L54) is so generic.

**Comment by ToucheSir** [comment]
I would prefer to even ditch the trait and just check for a set of known good types like the ones you listed to determine if `x` can be indexed. Otherwise the overall plan sounds reasonable.

**Comment by ToucheSir** [comment]
> * the `copyto!(buffer, src)` adjoint had this line `grad .= eltype(grad) <: Number ? 0 : nothing` and I removed it but no tests failed. What does this do exactly?

I think it's for the case where you have a buffer of non-numbers. Examples would be a buffer of differentiable structs, or a buffer of arrays. That neither of these cases were tested is bad, but also not uncommon for Zygote (which historically has poor test coverage in general).

> * The `copyto!` adjoint w.r.t. buffer is always nothing, but I suppose this is the intended behavior of the buffer? I would have tried the following but it seems prohibited: ...

My understanding is that for correctness reasons, differentiable arguments which are mutated should not have gradients returned for correctness reasons. Instead, a copy is kept in the mutable gradient cache managed by `grad_mut` and returned at either the top level or the point at which the mutable value was constructed.

File: src/lib/array.jl [hunk 1]
```diff
@@ -296,6 +297,20 @@ end
...
+  take_pullback(dy::NamedTuple{(:n,:xs)}) = (dy.xs, dy.n)
+  take_pullback(dy) = (takefunc(itr, dy), nothing)
+  Iterators.take(itr, n), take_pullback
+end
+
+@adjoint function Base.collect(t::Iterators.Take)
+    collect_take_pullback(dy) = ((xs=takefunc(t.xs, dy), n=nothing),)
+    collect(t), collect_take_pullback
+end
+
```
[the remark this answers is not in the document]
Comment (ToucheSir) [reply src/lib/array.jl:313]: That works, though if it's not too difficult to add tests for the `Iterators.take` rule itself we might as well get the 2-for-1 deal here?

**Review by ToucheSir — APPROVED** [review]
Thanks again for implementing a pretty tricky feature, @lxvm !

**Pull request merged**